### PR TITLE
Quimb Tensor Network Simulators

### PR DESCRIPTION
**Pull request by mpharrigan**
[Quimb](https://quimb.readthedocs.io/en/latest/index.html) is a nice tensor network package. This set of utilities (in `contrib` because there are features missing) allows you to turn cirq Circuits into quimb tensor networks with some benefits.

image

image

The particular use cases I've explored explicitly is

1. For shallow, low degree circuits you can compute the value of low weight observables on a very high number of qubits (example provided in "Contract a Grid Circuit" notebook)
2. For low-qubit number, deep, noisy circuits it is faster than the cirq density matrix simulator

image

## Features

 - Turn circuits into tensor networks representing state vector or unitary
 - Drop-in replacements for `cirq.final_state_vector` and `cirq.unitary` that use tensor contraction under the hood
 - Utilities for contracting circuits of the form <0 | U^dag A U |0> for PauliString `A`. 
 - Some related functionality for building circuits on a grid and simplifying some of the operations with low weight observables
 - Circuit-like positioning of tensors in tensor drawings for density matrix networks

## Limitations

 - Code duplication between state vector and density transforms
 - Initial state must be |0>
 - No circuit-like positioning of tensors in state-vector networks
 - No expectation values with density matrices
 - Doesn't follow the simulator and sampler APIs

### CODE REVIEW

File: cirq/contrib/quimb/density_matrix.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+    This assumes you start in the |0..0><0..| state. Indices are named
+    "nf{i}_q{x}" and "nb{i}_q{x}" where i is a time index and x is a
+    qubit index. nf- and nb- refer to the "forwards" and "backwards"
+    copies of the circuit. Kraus indices are named "k{j}" where j is an
+    independent "kraus" internal index which you probably never need to access.
+
+    Args:
+        circuit: The circuit containing operations that support the
+            cirq.unitary() or cirq.channel() protocols.
+        qubits: The qubits in the circuit.
```
Comment (Strilanc) [inline cirq/contrib/quimb/density_matrix.py:77]: Can this be an optional argument, defaulting to `sorted(circuit.all_qubits())`?

File: cirq/contrib/quimb/density_matrix.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+                               }))
+                kraus_frontier += 1
+            else:
+                raise ValueError(repr(op))  # coverage: ignore
+
+            _positions(mi + 1, op.qubits)
+    return tensors, qubit_frontier, positions
+
+
+def tensor_density_matrix(circuit, qubits):
```
Comment (Strilanc) [inline cirq/contrib/quimb/density_matrix.py:173]: Type annotations

File: cirq/contrib/quimb/density_matrix.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+        yb_offset: Offset the "backwards" circuit by this much.
+    """
+    qy = _qpos_y(qubits, n_qubits)
+    positions[(f'i{mi}f', _qpos_tag(qubits))] = \
+        (mi * x_scale + qy * x_nudge, y_scale * qy)
+    positions[(f'i{mi}b', _qpos_tag(qubits))] = \
+        (mi * x_scale, y_scale * qy + yb_offset)
+
+
+def circuit_to_density_matrix_tensors(circuit: cirq.Circuit,
```
Comment (Strilanc) [inline cirq/contrib/quimb/density_matrix.py:63]: Type annotation for result

File: cirq/contrib/quimb/state_vector.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
+from typing import Sequence, Union, List
+
+import quimb
+import quimb.tensor as qtn
+
+import cirq
+
+
+def circuit_to_tensors(circuit: cirq.Circuit,
```
Comment (Strilanc) [inline cirq/contrib/quimb/state_vector.py:9]: Result type annotation

File: cirq/contrib/quimb/state_vector.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+            U = cirq.unitary(op).reshape((2,) * 2 * len(op.qubits))
+            t = qtn.Tensor(data=U,
+                           inds=end_inds + start_inds,
+                           tags={f'Q{len(op.qubits)}'})
+            tensors.append(t)
+
+    return tensors, qubit_frontier, positions
+
+
+def tensor_state_vector(circuit, qubits):
```
Comment (Strilanc) [inline cirq/contrib/quimb/state_vector.py:71]: type annotations

File: cirq/contrib/quimb/state_vector.py [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+    """
+    tensors, qubit_frontier, _ = circuit_to_tensors(circuit=circuit,
+                                                    qubits=qubits)
+    tn = qtn.TensorNetwork(tensors)
+    f_inds = tuple(f'i{qubit_frontier[q]}_q{q}' for q in qubits)
+    tn.contract(inplace=True)
+    return tn.to_dense(f_inds)
+
+
+def tensor_unitary(circuit, qubits):
```
Comment (Strilanc) [inline cirq/contrib/quimb/state_vector.py:82]: type annotations

File: cirq/contrib/quimb/state_vector.py [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+                                                    qubits=qubits,
+                                                    initial_state=None)
+    tn = qtn.TensorNetwork(tensors)
+    i_inds = tuple(f'i0_q{q}' for q in qubits)
+    f_inds = tuple(f'i{qubit_frontier[q]}_q{q}' for q in qubits)
+    tn.contract(inplace=True)
+    return tn.to_dense(f_inds, i_inds)
+
+
+def circuit_for_expectation_value(circuit: cirq.Circuit,
```
Comment (Strilanc) [inline cirq/contrib/quimb/state_vector.py:95]: result type annotation

File: cirq/contrib/quimb/Contract-a-Grid-Circuit.ipynb [hunk 8]
```diff
@@ -0,0 +1,405 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%matplotlib inline\n",
+    "from matplotlib import pyplot as plt\n",
+    "\n",
+    "import seaborn as sns\n",
```
Comment (balopat) [inline cirq/contrib/quimb/Contract-a-Grid-Circuit.ipynb:44]: seaborn is missing from the contrib requirements - maybe worth adding it or at least mentioning it?

File: cirq/contrib/quimb/density_matrix.py [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+    if isinstance(qubits, cirq.LineQubit):
+        return _qpos_y((qubits,), n_qubits)
+    x = np.mean([q.x for q in qubits]).item()
+    return n_qubits - x - 1
+
+
+def _add_to_positions(positions: Dict[Tuple[str, str], Tuple[float, float]],
+                      mi: int,
+                      qubits: Union[cirq.LineQubit, Tuple[cirq.LineQubit]], *,
+                      n_qubits: int, x_scale, y_scale, x_nudge, yb_offset):
```
Comment (balopat) [inline cirq/contrib/quimb/density_matrix.py:40]: isn't n_qubits == len(qubits)?

File: cirq/contrib/quimb/density_matrix.py [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+    """
+    if isinstance(qubits, cirq.LineQubit):
+        return _qpos_tag((qubits,))
+    x = min(q.x for q in qubits)
+    return f'q{x}'
+
+
+@lru_cache()
+def _qpos_y(qubits: Union[cirq.LineQubit, Tuple[cirq.LineQubit]],
+            n_qubits: int):
```
Comment (balopat) [inline cirq/contrib/quimb/density_matrix.py:25]: isn't n_qubits == len(qubits)?

File: cirq/contrib/quimb/state_vector.py [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+            which case the tensor network will represent the final
+            state vector; or `None` in which case the starting indices
+            will be left open and the tensor network will represent the
+            circuit unitary.
+    Returns:
+        tensors: A list of quimb Tensor objects
+        qubit_frontier: A mapping from qubit to time index at the end of
+            the circuit. This can be used to deduce the names of the free
+            tensor indices.
+        positions: Currently None. May be changed in the future to return
```
[the remark this answers is not in the document]
Comment (mpharrigan) [reply cirq/contrib/quimb/state_vector.py:31]: I'm going to push back for the following three reasons

1. currently lets you switch between `circuit_to_tensors` and `circuit_to_density_matrix_tensors` with no other code changes
2. quimb utilities consuming this result handle `None` gracefully
3. I'm pretty sure this is going to get implemented, and changing return values is hard to do 
4. This function is a little lower-level than `tensor_unitary` and `tensor_state_vector`, and as such the user can probably handle dealing with the more "under-the-hood" tensors and qubit frontier (and positions)

File: cirq/contrib/quimb/Cirq-to-Tensor-Networks.ipynb [hunk 12]
```diff
@@ -0,0 +1,364 @@
...
+    "cirq.DropEmptyMoments().optimize_circuit(circuit)\n",
+    "SVGCircuit(circuit)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Circuit to Tensors\n",
+    "The circuit defines a tensor network representation. By default, the initial state is the `|0...0>` state (represented by the \"zero qubit\" operations labeled \"Q0\" in the legend. \"Q1\" are single qubit operations and \"Q2\" are two qubit operations. The open legs are the indexes into the state vector and are of the form \"i{m}_q{n}\" where `m` is the time index (given by the returned `qubit_frontier` dictionary) and \"n\" is the qubit string."
```
[the remark this answers is not in the document]
Comment (mpharrigan) [reply cirq/contrib/quimb/Cirq-to-Tensor-Networks.ipynb:66]: done! (locally)

File: cirq/contrib/quimb/density_matrix.py (hunk 10 above)
Comment (mpharrigan) [reply cirq/contrib/quimb/density_matrix.py:25]: No. Changing `n_qubits` to `tot_n_qubits`, the total number of qubits (needed to turn qubit index into a y position (which are backwards)). `qubits` is usually one or two qubits, corresponding to a tensor, corresponding to a gate operation

Comment (mpharrigan) [reply cirq/contrib/quimb/density_matrix.py:25]: and adding a docstring

File: cirq/contrib/quimb/density_matrix.py (hunk 9 above)
Comment (mpharrigan) [reply cirq/contrib/quimb/density_matrix.py:40]: see other comment; changed name; updated docstring

File: cirq/contrib/quimb/density_matrix.py (hunk 3 above)
Comment (mpharrigan) [reply cirq/contrib/quimb/density_matrix.py:63]: added

File: cirq/contrib/quimb/density_matrix.py [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+    positions[(f'i{mi}b', _qpos_tag(qubits))] = \
+        (mi * x_scale, y_scale * qy + yb_offset)
+
+
+def circuit_to_density_matrix_tensors(circuit: cirq.Circuit,
+                                      qubits: Sequence[cirq.LineQubit]):
+    """Given a circuit with mixtures or channels, construct a tensor network
+    representation of the density matrix.
+
+    This assumes you start in the |0..0><0..| state. Indices are named
```
[the remark this answers is not in the document]
Comment (mpharrigan) [reply cirq/contrib/quimb/density_matrix.py:68]: fixed (locally!). Thanks for the spot

File: cirq/contrib/quimb/density_matrix.py (hunk 1 above)
Comment (mpharrigan) [reply cirq/contrib/quimb/density_matrix.py:77]: done here and below with test

File: cirq/contrib/quimb/density_matrix.py (hunk 2 above)
Comment (mpharrigan) [reply cirq/contrib/quimb/density_matrix.py:173]: added

File: cirq/contrib/quimb/state_vector.py (hunk 5 above)
Comment (mpharrigan) [reply cirq/contrib/quimb/state_vector.py:71]: added

File: cirq/contrib/quimb/state_vector.py (hunk 6 above)
Comment (mpharrigan) [reply cirq/contrib/quimb/state_vector.py:82]: added

**Comment by mpharrigan** [comment]
I addressed all the comments and the build is green. I see you both gave approval, but I'll give a couple hours for any follow up reviews before I merge

File: cirq/contrib/quimb/state_vector.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (balopat) [reply cirq/contrib/quimb/state_vector.py:33]: Sounds good, you convinced me

**Comment by balopat** [comment]
From my end it looks good!

**Pull request merged**